src/vm/jitinterface.cpp

```diff
@@ -8912,6 +8912,72 @@ void CEEInfo::getMethodVTableOffset (CORINFO_METHOD_HANDLE methodHnd,
 }
 
 /*********************************************************************/
+CORINFO_METHOD_HANDLE CEEInfo::resolveVirtualMethod(CORINFO_METHOD_HANDLE methodHnd,
+                                                    CORINFO_CLASS_HANDLE derivedClass)
+{
+    CORINFO_METHOD_HANDLE result;
+
+    JIT_TO_EE_TRANSITION();
+
+    result = resolveVirtualMethodExact(methodHnd, derivedClass);
+
+    EE_TO_JIT_TRANSITION();
+
+    return result;
+}
+
+/*********************************************************************/
+CORINFO_METHOD_HANDLE CEEInfo::resolveVirtualMethodExact(CORINFO_METHOD_HANDLE methodHnd,
+                                                         CORINFO_CLASS_HANDLE derivedClass)
+{
+    CONTRACTL {
+        SO_TOLERANT;
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    CORINFO_METHOD_HANDLE result;
+
+    JIT_TO_EE_TRANSITION();
+
+    MethodDesc* method = GetMethod(methodHnd);
+
+    // Method better be from a fully loaded class
+    _ASSERTE(method->IsRestored() && method->GetMethodTable()->IsFullyLoaded());
+
+    // Method better be virtual
+    _ASSERTE(method->IsVirtual());
+
+    //@GENERICS: shouldn't be doing this for instantiated methods as they live elsewhere
+    _ASSERTE(!method->HasMethodInstantiation());
+
+    // Method's class better not be an interface
+    _ASSERTE(!method->GetMethodTable()->IsInterface());
+
+    // Method better be in the vtable
+    WORD slot = method->GetSlot();
+    _ASSERTE(slot < method->GetMethodTable()->GetNumVirtuals());
+
+    // TODO: Derived class should have base class as parent...
+    TypeHandle DerivedClsHnd(derivedClass);
+    MethodTable* pMT = DerivedClsHnd.GetMethodTable();
+
+    // MethodDescs returned to JIT at runtime are always fully loaded. Verify that it is the case.
+    _ASSERTE(pMT->IsRestored() && pMT->IsFullyLoaded());
+
+    MethodDesc* pMD = pMT->GetMethodDescForSlot(slot);
```

Comment:
canInline should work, but it is too pessimistic check and makes the optimization order dependent because of the noinlining hint is cached. I do not think you should be using canInline to gate it.